File: benchmarks/nyc_taxi/r/q10_vendor_revenue_pivot.R
args <- commandArgs(trailingOnly = TRUE)
if (length(args) < 1) {
  stop("usage: Rscript q10_vendor_revenue_pivot.R <parquet-dataset-path>", call. = FALSE)
}

suppressPackageStartupMessages({
  library(arrow)
  library(dplyr)
})

started <- Sys.time()
dataset_path <- args[[1]]
ds_tbl <- open_dataset(dataset_path, format = "parquet")

grouped <- ds_tbl |>
  group_by(year, month, VendorID) |>
  summarise(rev = sum(total_amount, na.rm = TRUE)) |>
  collect() |>
  mutate(vendor_label = ifelse(is.na(VendorID), "vendor_NA", paste0("vendor_", as.character(VendorID))))

result <- reshape(
  as.data.frame(grouped[, c("year", "month", "vendor_label", "rev")]),
  idvar = c("year", "month"),
  timevar = "vendor_label",
  direction = "wide"
)

print(head(result, 10))
cat("ROWS_SCANNED=NA\n")
cat(sprintf("ROWS_RETURNED=%s\n", nrow(result)))
cat(sprintf("ELAPSED_SEC=%.6f\n", as.numeric(difftime(Sys.time(), started, units = "secs"))))

File: benchmarks/nyc_taxi/python/q10_vendor_revenue_pivot.py
import argparse
import time

import pandas as pd
import pyarrow.dataset as ds


def vendor_label(value: object) -> str:
    if pd.isna(value):
        return "vendor_NA"
    if isinstance(value, float) and value.is_integer():
        return f"vendor_{int(value)}"
    return f"vendor_{value}"


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--dataset", required=True, help="Path to the partitioned parquet dataset")
    args = parser.parse_args()

    started = time.perf_counter()
    dataset = ds.dataset(args.dataset, format="parquet", partitioning="hive")

    table = dataset.to_table(columns=["year", "month", "VendorID", "total_amount"])
    frame = table.to_pandas()
    grouped = (
        frame.groupby(["year", "month", "VendorID"], dropna=False)
        .agg(rev=("total_amount", "sum"))
        .reset_index()
    )
    grouped["vendor_label"] = grouped["VendorID"].map(vendor_label)
    result = (
        grouped.pivot(index=["year", "month"], columns="vendor_label", values="rev")
        .reset_index()
        .sort_values(["year", "month"])
    )

    print(result.head(10))
    print("ROWS_SCANNED=NA")
    print(f"ROWS_RETURNED={len(result.index)}")
    print(f"ELAPSED_SEC={time.perf_counter() - started:.6f}")


if __name__ == "__main__":
    main()
